### PR TITLE
Unify 's' rules with 1990 paper

### PR DESCRIPTION
**Pull request by Hopper262**
First off, thanks for this module! I'm converting some code to JavaScript, and was pleased to find this less-prevalent stemmer was already on npm.

When testing my own code's results, I discovered the results don't quite match the rules in the [original paper](https://doi.org/10.1145/101306.101310), which match implementations including [NLTK](https://www.nltk.org/_modules/nltk/stem/lancaster.html) and [my own](https://github.com/Hopper262/paice-husk-stemmer). There are two rules which are off:

* [line 119](https://github.com/words/lancaster-stemmer/blob/6ddc40659afef59c94875d396da743eebe2e4788/index.js#L119) uses type `cont`, but the paper's rules also set the `intact` flag. Without that, the subsequent rule is never accessed, so that doesn't seem right.
* [line 120](https://github.com/words/lancaster-stemmer/blob/6ddc40659afef59c94875d396da743eebe2e4788/index.js#L120) uses a blank replacement, but the paper uses 's' (zero characters removed).

I assume both of these are due to starting from the C/Perl implementations posted on the Lancaster site, rather than the Pascal/Java implementations. The one-off "contint" type instead of "continue" is easy to miss, and the stop rule is flat-out different. I suspect this latter change was a mistake introduced during the rule format conversion, rather than a deliberate evolution, but I don't have proof beyond a (long-lost and possibly misremembered) brief exchange with Dr. Paice years ago.

I'm not really a researcher and can't say whether these changes improve the stemming, but this PR improves compatibility with the other readily-available implementations FWIW. I've also added a new test case with an affected word ("compensation").

### CODE REVIEW

**Comment by wooorm** [comment]
Hi! Interesting?

Why is paper/Pascal/Java/NLTK preferred over C/Perl?

As you say “PR improves compatibility with the other readily-available implementations FWIW”, but if there are two different versions in the wild, changing this project from one to the other *changes* compatibility from one group to another, but not really *improves* it.

**Comment by wooorm** [comment]
I just pushed some commits to fix the build and improve some tooling for the repo!

**Comment by Hopper262** [comment]
Well, all implementations including C/Perl use the `contint` type for line 119, so this implementation is unique from all others in the wild in that respect. It was probably a mistake to bring line 120 into the same PR and muddy the issue. I can update the PR to narrow the focus.

Would you be interested in a feature PR to allow for replacement rulesets? That would make the "defaults" much less important, as anyone with specific preferences or compatibility needs could just provide custom rules instead of forking the project.

**Comment by wooorm** [comment]
Then `contint` sounds good, and a bug fix!

> Would you be interested in a feature PR to allow for replacement rulesets?

Are there more (intentional, not bug) differences between projects?
Is there an actual benefit of allowing arbitrary, user-provided rules, or would we reasonably just have a couple of small differences in rules?

If this is one small difference, then I’d prefer a more simple boolean option to switch between the default and a different flavor?

**Comment by Hopper262** [comment]
If the main goal is simplicity, adding "export" to "const rules = ..." would expose the rules object and allow modifications without affecting the rest of the code or documented API. However, I found some surprising stuff while comparing the existing rules, and a switch might be more useful overall.

In searching for other implementations and/or rules, I came across [Whoosh](https://github.com/mchaput/whoosh/blob/main/src/whoosh/lang/paicehusk.py), which strips a fixed set of prefixes but otherwise doesn't introduce anything new. The only true customization of the rules I found is in [Antonio Zamora's paper](https://www.scientificpsychic.com/paice/paice.html), which is a major overhaul that also includes at least one algorithm change (to minimum stem size). If there is a community of rule-tinkerers, they're pretty well hidden.

* Paper / Pascal / Java / NLTK / Hopper / Whoosh share rules. Comparing the comments to the rules, there are three discrepancies, and they all appear to be rule errors where the replacement text was not reversed:
    * `hsiug5ct.` should be `hsiug5tc.` ( -guish > -ct )
    * `noix4ct.` should be `noix4tc.` ( -xion > -ct )
    * `tpec2iv.` should be `tpec2vi.` ( -cept > -ceiv )
* C / Perl share rules, which fix the three rules above and introduce one other change. Comparing the comments to the rules, there were three lines with missing hyphens and three other lines with discrepancies:
    * `s,?,protect` replaced with `s,?,stop` ( -s > -- )
    * `i,y,continue` commented as ( -i > -y ) instead of ( -i > -y cont )
    * `ifiabl,?,stop` commented as ( -ifiable > -- ) instead of (-ifiabl > -- )
    * `rr,r,stop` commented as the misformatted ( -rr > - ) instead of ( -rr > -r )
* This module has two rule differences from C / Perl:
    * `s,?,contint` replaced with `s,?,continue`
    * `ior,?,continue` rule added

Note: I only compared the rulesets, not the actual code; the real-world compatibility situation might be worse. Even so, documenting these rule variants in code might be a useful service for future researchers.

When I started this conversation, I wasn't expecting to find bugs in the paper's rules and consider revising the stemmer used at my day job. Thanks for asking questions and giving me an excuse to do some digital archaeology 😄

**Comment by wooorm** [comment]
> adding "export" to "const rules = ..."

This might be simplest for us here, but I feel like that isn’t simple for users, and it feels to me like leaking internals if users are supposed to mutate those objects, I’d prefer one or two options to change to different behavior instead.
In my experience with the Node.js ecosystem, it’s also generally not a good idea to do that: perhaps some other package might use this project too, a user meddling with the package here, would also change how that project would work. Potentially breaking it. (Or not, depending on how npm installs packages)

> strips a fixed set of prefixes but otherwise doesn't introduce anything new

If someone would want that, they could wrap this project and implement those same rules outside, correct?

> which is a major overhaul that also includes at least one algorithm change

That sounds like a very different algo, so I don’t think project needs to concern itself with that!

---

Could you summarize for me which of the two, L119 and L120, was the bug that is clearly wrong? And which one was a difference between paper/Pascal/Java/NLTK and C/Perl?
For the bug fix, I think we can just have that.
For the difference, I’d prefer an option if possible?

**Comment by Hopper262** [comment]
L119 is the bug, L120 is the difference. I'll update this PR to make L120 an option, with the default being the current behavior. Also, L104 (-ior) is not present in either of the original rulesets. Should I make that an option, or remove it?

My original goal for this PR was to improve interoperability between codebases, but after [testing the original implementations](https://github.com/Hopper262/paice-husk-stemmer/blob/master/compatibility.md), I realize that's a much bigger and messier project than I anticipated. The option will clarify which rules are being used here, and that's good enough for now IMO.

**Comment by Hopper262** [comment]
OK, this is ready for feedback now. Default usage does not change. The paper's rules can be requested with an optional second parameter:

`lancasterStemmer('word', { ruleset: '1990' });`

I also included a "2014" ruleset, which provides the rules used before this PR, but it's easy to remove if not necessary.

I used JSON for object deep cloning, as structuredClone() isn't in Node 16 and a runtime dependency seems like overkill just for this. I could also avoid the cloning at the expense of duplicating the rule definitions. If you'd like to see this done differently, or have any other style or substance requests, let me know.

**Comment by wooorm** [comment]
Thanks! I made a branch that’s almost done: https://github.com/words/lancaster-stemmer/compare/algos-1990-1994. But I had some Qs for you first!

What I did:

- refactored the code
- dropped 2014
- switched to define 1990 and mutate 1994
- added a bunch of docs

Couple of Qs:

- Which is the better default? Do you see 1994 as an anomaly, a bug? Or is 1990 the bug?
- Should we use years? Because some of the implementations that match 1990 are made later, and the same for 1994.

Could you check if you agree with the docs? Whether everything else looks good?

If everything’s good, I’ll still add this as an option to the CLI

**Comment by wooorm** [comment]
Ping! :)

**Comment by Hopper262** [comment]
Haha, thanks for the ping. I was headed out of town right as this came in, then I had work to catch up on after I got back, and the weeks just slid past. Sorry for disappearing!

This is awesome, excellent work! I really appreciate you taking my attempt and improving the TypeScript and documentation parts I was unsure about -- I'm very new to TS so it was instructive to see how you handled the options.

* For the default, I lean slightly toward 1990. I'm not sure that 1994 is a bug, as I can understand the reasoning behind that change, but it is an anomaly since the 2000 Java release used the original 1990 rules. There are other errors and sloppiness in the comments of the 1994 rules, whereas the 1990 ones look clean. (I mentioned a bug in the 1990 rules earlier in this thread, but that was due to my misreading the rule format -- totally my mistake and not Paice's.)
* Having mulled this over, I agree years aren't the best. Something like "paper" and "c"/"ansic" would be more obviously tied to the sources.

For the docs, I like the use of the more generic "style" as the option, but we could be clearer that there are non-ruleset differences (for example, when to stop stemming) which aren't being copied here. Maybe something like this:

> There are small algorithmic differences between how the algorithm was
> implemented over the years.
> Looking at [Algorithm Implementations][algos] on the archived website,
> there are four styles available, in addition to the original paper.
> 
> The only difference currently implemented in this package is whether a final
> `s` is kept before stopping ("paper") or dropped before stopping ("c").

Thank you again for indulging in my research and digressions, and for building a lovely code base that was a joy to work with!

**Comment by wooorm** [comment]
Thanks, released in https://github.com/words/lancaster-stemmer/releases/tag/2.1.0!
Kept the default at how it’s currently is. It’s probably marginal but it might be classified as a breaking change otherwise
